examples/step-57/step-57.cc

```diff
@@ -0,0 +1,948 @@
+  // in the second factor is simple matrix-vector multiplication. The Schur complement $\tilde{S}$
+  // can be well approximated by the mass matrix of pressure and its inverse can be obtained through an inexact
+  // solver. Because of the symmetry of pressure mass matrix we use an iterative method, the CG method, to
+  // solve the corresponding linear system.
+  //
+  // In summary, the preconditioner is defined by a combination of operators: a direct solver for the first
+  // factor, matrix-vector multiplication for the second factor and a iterative method for the last factor.
+  // gamma is the coefficient of the Augmented Lagrangian term.
+
+
+  template <class PreconditionerMp>
+  class BlockSchurPreconditioner : public Subscriptor
+  {
+  public:
+    BlockSchurPreconditioner (double                                     gamma,
+                              double                                     viscosity,
+                              const BlockSparseMatrix<double>            &S,
+                              const SparseMatrix<double>                 &P,
+                              const PreconditionerMp                     &Mppreconditioner
+                             );
+
+    void vmult (BlockVector<double>       &dst,
+                const BlockVector<double> &src) const;
+
+  private:
+    const double gamma;
+    const double viscosity;
+    const BlockSparseMatrix<double> &stokes_matrix;
+    const SparseMatrix<double>      &pressure_mass_matrix;
+    const PreconditionerMp          &mp_preconditioner;
+  };
+
+  template <class PreconditionerMp>
+  BlockSchurPreconditioner<PreconditionerMp>::
+  BlockSchurPreconditioner (double                           gamma,
+                            double                           viscosity,
+                            const BlockSparseMatrix<double>  &S,
+                            const SparseMatrix<double>       &P,
+                            const PreconditionerMp           &Mppreconditioner)
+    :
+    gamma                (gamma),
+    viscosity            (viscosity),
+    stokes_matrix        (S),
+    pressure_mass_matrix (P),
+    mp_preconditioner    (Mppreconditioner)
+  {}
+
+  template <class PreconditionerMp>
+  void
+  BlockSchurPreconditioner<PreconditionerMp>::
+  vmult (BlockVector<double>       &dst,
+         const BlockVector<double> &src) const
+  {
+    Vector<double> utmp(src.block(0));
+
+    {
+      SolverControl solver_control(1000, 1e-6 * src.block(1).l2_norm());
+      SolverCG<>    cg (solver_control);
+
+      dst.block(1) = 0.0;
+      cg.solve(pressure_mass_matrix,
+               dst.block(1), src.block(1),
+               mp_preconditioner);
+      dst.block(1) *= -(viscosity+gamma);
+    }
+
+    {
+      stokes_matrix.block(0,1).vmult(utmp, dst.block(1));
+      utmp*=-1.0;
+      utmp+=src.block(0);
+    }
+
+    SparseDirectUMFPACK  A_direct;
+    A_direct.initialize(stokes_matrix.block(0,0));
```

Comment:
Don't you want to move the factorization outside of the preconditioner class and hand it in similarly to the ILU? Or rather, even move it into the general NavierStokesNewton class where it can be kept over different nonlinear iterations? It seems that you only assemble the system matrix in the first nonlinear step so you could re-use it otherwise.